source/commoncrypto/commoncrypto_ecc.c

```diff
@@ -0,0 +1,607 @@
+        default:
+            return 0;
+    }
+}
+
+static uint8_t s_preamble[] = {
+    0x04,
+};
+
+static size_t s_der_overhead = 8;
+
+static int s_sign_message_fn(
+    const struct aws_ecc_key_pair *key_pair,
+    const struct aws_byte_cursor *message,
+    struct aws_byte_buf *signature_output) {
+    struct commoncrypto_ecc_key_pair *cc_key = key_pair->impl;
+
+    if (!cc_key->priv_key_ref) {
+        return aws_raise_error(AWS_ERROR_CAL_MISSING_REQUIRED_KEY_COMPONENT);
+    }
+
+    CFDataRef hash_ref =
+        CFDataCreateWithBytesNoCopy(cc_key->cf_allocator, message->ptr, message->len, kCFAllocatorNull);
+
+    CFErrorRef error = NULL;
+    CFDataRef signature =
+        SecKeyCreateSignature(cc_key->priv_key_ref, kSecKeyAlgorithmECDSASignatureDigestX962, hash_ref, &error);
+
+    struct aws_byte_cursor to_write =
+        aws_byte_cursor_from_array(CFDataGetBytePtr(signature), CFDataGetLength(signature));
+    aws_byte_buf_append(signature_output, &to_write);
+
+    CFRelease(signature);
+    CFRelease(hash_ref);
+
+    (void)error;
+    return AWS_OP_SUCCESS;
+}
+
+static size_t s_signature_length_fn(const struct aws_ecc_key_pair *key_pair) {
+    return s_key_coordinate_byte_size_from_curve_name(key_pair->curve_name) * 2 + s_der_overhead;
+}
+
+static int s_verify_signature_fn(
+    const struct aws_ecc_key_pair *key_pair,
+    const struct aws_byte_cursor *message,
+    const struct aws_byte_cursor *signature) {
+    struct commoncrypto_ecc_key_pair *cc_key = key_pair->impl;
+
+    if (!cc_key->pub_key_ref) {
+        return aws_raise_error(AWS_ERROR_CAL_MISSING_REQUIRED_KEY_COMPONENT);
+    }
+
+    CFDataRef hash_ref =
+        CFDataCreateWithBytesNoCopy(cc_key->cf_allocator, message->ptr, message->len, kCFAllocatorNull);
+    CFDataRef signature_ref =
+        CFDataCreateWithBytesNoCopy(cc_key->cf_allocator, signature->ptr, signature->len, kCFAllocatorNull);
+
+    CFErrorRef error = NULL;
+
+    bool verified = SecKeyVerifySignature(
+        cc_key->pub_key_ref, kSecKeyAlgorithmECDSASignatureDigestX962, hash_ref, signature_ref, &error);
+
+    CFRelease(signature_ref);
+    CFRelease(hash_ref);
+
+    return verified ? AWS_OP_SUCCESS : aws_raise_error(AWS_ERROR_CAL_SIGNATURE_VALIDATION_FAILED);
+}
+
+static int s_derive_public_key_fn(struct aws_ecc_key_pair *key_pair) {
+    /* we already have a public key, just lie and tell them we succeeded */
+    if (key_pair->pub_x.buffer && key_pair->pub_x.len) {
+        return AWS_OP_SUCCESS;
+    }
+
+    return aws_raise_error(AWS_ERROR_UNSUPPORTED_OPERATION);
+}
+
+static void s_destroy_key_fn(struct aws_ecc_key_pair *key_pair) {
+    struct commoncrypto_ecc_key_pair *cc_key = key_pair->impl;
+
+    if (cc_key->pub_key_ref) {
+        CFRelease(cc_key->pub_key_ref);
+    }
+
+    if (cc_key->priv_key_ref) {
+        CFRelease(cc_key->priv_key_ref);
+    }
+
+    aws_byte_buf_clean_up_secure(&key_pair->key_buf);
+    aws_mem_release(key_pair->allocator, cc_key);
+}
```

Comment:
need to clean up `cc_key_pair->cf_allocator`

Reply:
good catch.